### DISCUSSION TITLE
Telnet Connection Unsuccessful - Exception?

### DISCUSSION THREAD

**Discussion by rgourleypython** [ktbyers/netmiko#2299] [Q&A] [answered]
I am attempting to determine if a telnet connection was unsuccessful due to the host being offline or unreachable so I can attempt a ping afterwards. I am using the generic_telnet device type to connect to Windows XP systems. For systems that are reachable, the connection is not being established automatically so I am having to use find_prompt() and write_channel() to detect the prompt and send the username and password, which is working fine. However, if a system is offline, it fails to connect, but doesn't return any exception. Is there a way to determine if the connection was unsuccessful? Neither the NetmikoAuthenticationException nor the NetmikoTimeoutException are being thrown. I have the same issue with version 3.3.3 and 3.4.

Thank you,
Ryan

**Comment by ktbyers** [comment]
Have you tried passing in a lower timeout, maybe `timeout=10`?

**Reply by rgourleypython** [reply]
Just tried that. Same thing happens, it doesn't connect, but doesn't throw an exception either. I think it might be related to having to manually send the username and password instead of the telnet_login() function doing it, which appears to be the function where the exceptions can be thrown. Any idea why it's not connecting automatically? It appears that it may be because the telnet_login() function is using read_channel(), whereas I am having to use find_prompt() to find "login" and "password" before sending them. When I run read_channel() it returns "" or None, but perhaps read_channel() is consuming that within telnet_login() before I can read.  The username and password patterns in netmiko should be able to match from what I can tell.

**Reply by ktbyers** [reply]
Can you post your existing code?

**Reply by rgourleypython** [reply]
There are multiple python files involved, but I tried to simplify it here. I recently set auto_connect to False and am calling establish_connection() in the connect() function. The following are the 2 functions within a class that are related to doing the telnet connection. I loop through and create an object of the class for each host I need to connect to. If I get rid of my for loop in the connect() function, it immediately moves on, suggesting it's not even attempting the authentication through the telnet_login() function in netmiko. 

Also note this does work for systems that are reachable, it's just that I can't detect the ones that aren't reachable so I can try to ping them.

```
def __init__(self, ip: str, user: str, pw: str, script_setup_config: dict, log_obj=None):
        self.log_obj = log_obj
        self.file_timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
        self.script_path = Path(path.dirname(sys.argv[0]))
        self.device = {
            "device_type": "generic_telnet",
            "ip": ip,
            "host": ip,
            "username": user,
            "password": pw,
            "secret": pw,
            "port": 23,
            "global_delay_factor": 1,
            "fast_cli": False,
            "auto_connect": False,
            "timeout": 10
        }
        if script_setup_config.get("telnet_logs"):
            # set up telnet session logging
            makedirs(f"{self.script_path}\\logs\\{self.log_obj.script_name}\\telnet", exist_ok=True)
            self.device["session_log"] = f"{self.script_path}\\logs\\{self.log_obj.script_name}\\telnet\\{self.log_obj.script_name}_{ip}_{self.file_timestamp}.log"
        self.net_connect = netmiko.ConnectHandler(**self.device)
        self.connected = self.connect()

def connect(self):
    try:
        self.net_connect.establish_connection()
    except netmiko.NetmikoAuthenticationException as error:
        # this never happens
        print(f"Netmiko Authentication Exception: {error}")
        self.ping()
        return False
    except netmiko.NetmikoTimeoutException as error:
        # this never happens
        print(f"Netmiko Timeout Exception: {error}")
        self.ping()
        return False
    else:
        for _ in range(1, 10):
            sleep(self.net_connect.global_delay_factor)
            prompt = self.net_connect.find_prompt()
            if "login" in prompt:
                # print("found login")
                self.net_connect.write_channel(self.net_connect.username)
            elif "password" in prompt:
                # print("found pw")
                self.net_connect.write_channel(self.net_connect.password)
            elif '>' in prompt:
                return True
            else:
                # this never happens
                self.ping()
                return False

```

**Reply by ktbyers** [reply]
Why are you writing your own `connect()` instead of just using the standard Netmiko way of doing this.

You can do what you want, but I might not want to debug it for you.

**Reply by rgourleypython** [reply]
I would prefer to let Netmiko do it for me, but if I remove "auto_connect": False or set it to True, it does not connect. The telnet and Netmiko logs are empty in this case as well. I have to use find_prompt() to find "login" and "password" and use write_channel() to send the user and password manually or nothing happens.

**Reply by rgourleypython** [reply]
I do appreciate your quick responses to my question. Thank you! I may be missing something simple, but calling ConnectHandler should establish the connection if auto_connect hasn't been set to false, right? If so, something isn't working right or I am missing something.

**Comment by ktbyers** [comment]
@rgourleypython If you do the telnet manually to one of these systems what does the login process look like?

On quick glance, I think the issue is probably that you are using `generic_telnet` which requires you to handle the `telnet_login` process yourself and is generally intended for terminal servers.

**Reply by rgourleypython** [reply]
Process:

        telnet [IP]
        
        Welcome to Microsoft Telnet Service
        
        login: ***
        password: ***
        
        *===============================================================
        Microsoft Telnet Server.
        *===============================================================
        C:\Users\***>


Is there another device type I should use to telnet to Windows systems, particularly XP?

**Comment by ktbyers** [comment]
I would try this:

1. Set auto_connect = True

2. Disable your own .connect() call
3. Right after you call ConnectHandler try:

BaseConnection.telnet_login(self.net_connect)

You will have to import BaseConnection into your code i.e.

`from netmiko import BaseConnection`

Let me know if that improves things at all?

**Reply by ktbyers** [reply]
Basically, this is trying to invoke the `telnet_login` method that is in the BaseConnection class (to handle the telnet login process).

**Reply by rgourleypython** [reply]
That makes sense. This suggestion works for systems I can telnet to, which simplifies things for me to make the connection. It still won't return an exception for an unreachable system, for which the session and netmiko logs are empty. I also have a system that is stuck in a shutdown mode that returns "A system shutdown is in progress." after the password is entered, which is logged in the session and netmiko logs, but no exception there either. I understand that may be something I need to handle myself. Does Netmiko have a boolean or something I can check for whether it thinks the connection is established and authenticated?

**Comment by ktbyers** [comment, marked as answer]
Exceptions work for me as long as I specify the timeout. This is for a system that doesn't have telnet enabled:

```python
from netmiko import ConnectHandler, BaseConnection
from getpass import getpass
import logging

host = "cisco.domain.io"
username = "admin"
password = getpass()

net_connect = ConnectHandler(
    host=host, username=username, password=password, device_type="generic_telnet", timeout=5
)
BaseConnection.telnet_login(net_connect)
print("-------")
print(net_connect.find_prompt())
print("-------")
```

```
$ python netmiko_simple.py 
Password: 
Traceback (most recent call last):
  File "netmiko_simple.py", line 9, in <module>
    net_connect = ConnectHandler(
  File "/home/kbyers/netmiko/netmiko/ssh_dispatcher.py", line 330, in ConnectHandler
    return ConnectionClass(*args, **kwargs)
  File "/home/kbyers/netmiko/netmiko/base_connection.py", line 426, in __init__
    self._open()
  File "/home/kbyers/netmiko/netmiko/base_connection.py", line 431, in _open
    self.establish_connection()
  File "/home/kbyers/netmiko/netmiko/base_connection.py", line 913, in establish_connection
    self.remote_conn = telnetlib.Telnet(
  File "/usr/local/lib/python3.8/telnetlib.py", line 218, in __init__
    self.open(host, port, timeout)
  File "/usr/local/lib/python3.8/telnetlib.py", line 235, in open
    self.sock = socket.create_connection((host, port), timeout)
  File "/usr/local/lib/python3.8/socket.py", line 808, in create_connection
    raise err
  File "/usr/local/lib/python3.8/socket.py", line 796, in create_connection
    sock.connect(sa)
socket.timeout: timed out
```

**Reply by rgourleypython** [reply]
I tested this in a separate python file and the exception was thrown. I found the problem to be related to using `concurrent.futures.ThreadPoolExecutor` to create threads for each host connection. I can use `concurrent.futures.as_completed` to capture the exceptions so I should be able to take action based on that. I appreciate the help you have provided.

Thank you,
Ryan
